sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -510,6 +510,10 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         } else {
           e.copy(branches = branches.take(i).map(branch => (branch._1, elseValue)))
         }
+
+      case EqualTo(CaseWhen(branches, _), right)
```

Comment:
I'm a bit worried about dropping other branches in CASE WHEN. `a.semanticEquals(b)` means `a` is always equal to `b`. But `!a.semanticEquals(b)` doesn't mean that `a` will never be equal to `b`.

Reply:
As an example `(CASE WHEN a=1 THEN 1 ELSE b) = 1` can be true if `a=1` or `b=1`.